crates/bevy_math/src/gcd.rs

```diff
@@ -0,0 +1,158 @@
+//! Provides implementations for methods relating to the [Greatest Common Divisor](https://en.wikipedia.org/wiki/Greatest_common_divisor).
+
+/// Computes the [Greatest Common Divisor](https://en.wikipedia.org/wiki/Greatest_common_divisor)
+/// between the supplied `value` and the constant parameter `N`.
+///
+/// This method uses the [Euclidean Algorithm](https://en.wikipedia.org/wiki/Euclidean_algorithm#Implementations).
+#[inline(always)]
+pub const fn gcd(mut a: usize, mut b: usize) -> usize {
+    while b != 0 {
+        let t = b;
+        b = a % b;
+        a = t;
+    }
+    a
+}
+
+/// Computes the [Greatest Common Divisor](https://en.wikipedia.org/wiki/Greatest_common_divisor)
+/// between the supplied `value` and the constant parameter `N`.
+///
+/// This method uses a lookup table computed at compile-time, avoiding branches.
+/// If the second value `N` is not a constant knowable at compile-time, or it's
+/// impractically large, consider using [`gcd`] instead.
+///
+/// This method does _not_ panic.
+#[inline(always)]
+pub const fn gcd_by_table<const N: usize>(value: usize) -> usize {
+    const fn gcd_table<const N: usize>() -> [usize; N] {
+        let mut table: [usize; N] = [1; N];
+        let mut i = 0;
+
+        while i < N {
+            table[i] = gcd(N, i);
+
+            i += 1;
+        }
+
+        table
+    }
+
+    if N == 0 {
+        return 0;
+    }
+
+    let table = const {
+        // Taking a reference avoids copying this table
+        &gcd_table::<N>()
+    };
+
+    table[value % N]
+}
+
+/// Computes `N / gcd(N, value)`, where `gcd(N, value)` is the [Greatest Common Divisor](https://en.wikipedia.org/wiki/Greatest_common_divisor)
+/// between the supplied `value` and the constant parameter `N`.
+///
+/// This method uses a lookup table computed at compile-time, avoiding branches.
+/// If the second value `N` is not a constant knowable at compile-time, or it's
+/// impractically large, consider using [`gcd`] directly instead.
+///
+/// This method does _not_ panic.
+///
+/// For the case where `N` is zero, the returned value is zero, instead of a runtime panic.
+#[inline(always)]
+pub const fn n_over_gcd_by_table<const N: usize>(value: usize) -> usize {
+    const fn n_over_gcd_table<const N: usize>() -> [usize; N] {
+        let mut table: [usize; N] = [1; N];
+        let mut i = 0;
+
+        while i < N {
+            table[i] = N / gcd(N, i);
+
+            i += 1;
+        }
+
+        table
+    }
+
+    if N == 0 {
+        return 0;
+    }
+
+    let table = const {
+        // Taking a reference avoids copying this table
+        &n_over_gcd_table::<N>()
+    };
+
+    table[value % N]
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn gcd_tests() {
```

Comment:
Due to how the implementation works, I wouldn't take for granted that `gcd(a,b) = gcd(b,a)`, so an extra test with the reversed order of arguments would be nice imo. (at least add one for the a = 0, b != 0 case)

Reply:
That is a fair concern. I've added a test for symmetry, refactored the implementation of `gcd` to be a little cleaner, and cited where the specific implementation comes from, to hopefully alleviate further concerns around validity.